daemon/src/peer.rs

```diff
@@ -0,0 +1,288 @@
+                }
+            }
+        }
+    }
+
+    async fn run(mut self) {
+        loop {
+            tokio::select! {
+                _ = self.tcp_handle.shutdown_token.cancelled() => {
+                    debug!("Shutting down main start_sync loop");
+                    break;
+                }
+                Some(message) = self.syncer_receiver.recv() => {
+                    self.handle_message(message).await;
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct SyncActorHandle {
+    syncer_message_tx: SyncerMessageSender,
+}
+
+impl SyncActorHandle {
+    pub fn new(document_handle: DocumentActorHandle, tcp_handle: TCPActorHandle) -> Self {
+        let (syncer_message_tx, syncer_message_rx) = mpsc::channel(16);
+
+        // Sync actor.
+        let syncer = SyncActor::new(
+            syncer_message_rx,
+            document_handle.clone(),
+            tcp_handle.clone(),
+        );
+        tokio::spawn(syncer.run());
+
+        // Generate sync message when doc changes.
+        let shutdown_token_clone = tcp_handle.shutdown_token.clone();
+        let mut doc_changed_ping_rx = document_handle.subscribe_document_changes();
+        let syncer_message_tx_clone = syncer_message_tx.clone();
+
+        // TODO: can we explain here, why this forwarding is necessary?
+        tokio::spawn(async move {
+            loop {
+                syncer_message_tx_clone
+                    .send(SyncerMessage::GenerateSyncMessage {})
+                    .await
+                    .expect("Failed to send GenerateSyncMessage to document task");
+                tokio::select! {
+                    _ = shutdown_token_clone.cancelled() => {
+                        debug!("Stopping GenerateSyncMessage ping forwarding.");
+                        break;
+                    }
+                    doc_ping = doc_changed_ping_rx.recv() => match doc_ping {
+                        Ok(()) => {
+                            debug!("Doc changed.");
+                        }
+                        Err(broadcast::error::RecvError::Closed) => {
+                            panic!("Doc changed channel has been closed");
+                        }
+                        Err(broadcast::error::RecvError::Lagged(_)) => {
+                            // This is fine, the messages in this channel are just pings.
+                            // It's okay if we miss some.
+                        }
+                    }
+                }
+            }
+        });
+
+        Self { syncer_message_tx }
+    }
+
+    async fn send(&self, message: SyncerMessage) {
+        self.syncer_message_tx
+            .send(message)
+            .await
+            .expect("Channel closed (TODO)")
+    }
+}
+
+#[derive(Clone)]
+pub struct TCPActorHandle {
+    automerge_message_tx: mpsc::Sender<AutomergeSyncMessage>,
+    shutdown_token: CancellationToken,
+}
+
+/// The TCP statemachine works as follows:
```

Comment:
- [ ] finish/rework this docstring (and I think it's not in the right place anymore)